### PR TITLE
TINY-4926: Respect table_style_with_css when reading properties in the table dialog

### PR DESCRIPTION
**Pull request by ashfordneil**
Related Ticket: TINY-4926

Description of Changes:
* Small tweaks to how we read things in the table dialog
* A new test (this will probably need some work to fit in better with other table tests)

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] ~Branch prefixed with `feature/` for new features (if applicable)~

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable): N/A

### CODE REVIEW

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    Attribute.setAll(defaultTable, tableAttributes);
+    Css.setAll(defaultTable, tableStyles);
+
+    Arr.each(SelectorFilter.descendants(defaultTable, 'td,th'), (elem) => Css.setAll(elem, cellStyles));
+
+    editor.setContent(Html.getOuter(defaultTable));
+    TinySelections.setCursor(editor, [ 0, 0, 0 ], 0);
+  };
+
+  const assertTable = (editor: Editor, tableAttributes: Record<string, string>, tableStyles: Record<string, string>, cellStyles: Record<string, string>) => {
```
Comment (diogo405) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts:44]: I'd have:
`const assertTable = (editor: Editor, spec: { tableAttributes: Record<string, string>, tableStyles: Record<string, string>, cellStyles: Record<string, string> }) => {`

So in your tests would be easier to read:

`assertTable(editor, { tableAttributes: { cellspacing: '5px' }, tableStyles: { 'border-spacing': '5px' }, cellStyles: { padding: '30px' }});`

Same for `initializeTable`.

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        table_style_by_css: spec.style_by_css
+      }, [ Plugin, Theme ]);
+
+      it('falls back to cellpadding attribute if no CSS is defined', async () => {
+        const editor = hook.editor();
+        initializeTable(editor, { cellpadding: '20px' }, {}, {});
+
+        await TableTestUtils.pOpenTableDialog(editor);
+
+        TableTestUtils.assertDialogValues({
```
Comment (diogo405) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts:81]: I'd wrap `asserDialogValues` in a function and have inside the `false` and `generalSelectors` so you don't need to pass them every time as only the first param changes?!

**Review by diogo405 — APPROVED** [review]

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+          cellspacing: '5px',
+          cellpadding: 0
+        }, false, generalSelectors);
+
+        await TableTestUtils.pClickDialogButton(editor, true);
+
+        if (spec.style_by_css) {
+          assertTable(editor, { cellspacing: '' }, { 'border-spacing': '5px' }, {});
+        } else {
+          assertTable(editor, { cellspacing: '5px' }, { 'border-spacing': '' }, {});
```
Comment (ashfordneil) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts:195]: Apparently this sort of thing gets normalised to `5` on IE. How do we normally handle things like that in our tests?

Comment (lnewson) [reply modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts:195]: We set a valid value in the first place? `5px` is not a valid value for the cellspacing attribute on line 186 when using `table_style_by_css: false` as it needs to be a number not a CSS dimension.

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  ], (spec) => {
+    context(`Control table layout ${spec.title}`, () => {
+      const hook = TinyHooks.bddSetup<Editor>({
+        plugins: 'table',
+        base_url: '/project/tinymce/js/tinymce',
+        toolbar: 'tableprops',
+        table_style_by_css: spec.style_by_css
+      }, [ Plugin, Theme ]);
+
+      it('falls back to cellpadding attribute if no CSS is defined', async () => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts:75]: This probably doesn't apply to all tests, but you're missing the Jira prefix on some of these at least.

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+
+      return s.element('body', { children: [ table ] });
+    }));
+  };
+
+  Arr.each([
+    { title: 'using attributes', style_by_css: false },
+    { title: 'using styles', style_by_css: true }
+  ], (spec) => {
+    context(`Control table layout ${spec.title}`, () => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts:67]:
```suggestion
    context(`Table layout ${spec.title}`, () => {
```
This just feels really wrong to have `Control` here when you combine it with what's being tested.

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts (hunk 1 above)
Comment (ashfordneil) [reply modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts]: I went slightly further (compressed it all into one object with 4 optional fields) but this was a good catch, thanks

**Review by diogo405 — APPROVED** [review]

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts [hunk 6]
```diff
@@ -32,167 +43,184 @@ describe('browser.tinymce.plugins.table.ui.TableCellDialogTest', () => {
...
+      Arr.each(SelectorFilter.descendants(defaultTable, 'td,th'), (cell) => Css.set(cell, 'padding', table.cellPaddingStyle));
+    }
 
     editor.setContent(Html.getOuter(defaultTable));
     TinySelections.setCursor(editor, [ 0, 0, 0 ], 0);
   };
 
-  const assertTable = (editor: Editor, tableAttributes: Record<string, string>, tableStyles: Record<string, string>, cellStyles: Record<string, string>) => {
+  const assertTable = (editor: Editor, spec: TableSpec) => {
     TinyAssertions.assertContentStructure(editor, ApproxStructure.build((s, str, _arr) => {
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts:64]: Not related to this change and the `s, str, arr` is our common convention here.

File: modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+        assertDialogValues({
+          cellspacing: '',
+          cellpadding: '20'
+        });
+
+        await TableTestUtils.pClickDialogButton(editor, false);
+      });
+
+      it('TINY-4926: TINY-4926: falls back to td padding styles if no cellpadding attribute is defined', async () => {
```
Comment (lnewson) [inline modules/tinymce/src/plugins/table/test/ts/browser/ui/TableDialogStyleWithCssTest.ts:117]: Another duplicate prefix here.

**Review by lnewson — APPROVED** [review]

**Review by JamesToohey — APPROVED** [review]

**Pull request merged**